<hr>
<b><font color='red'>Ejercicio 01</font></b>  

Lee el archivo.txt 

In [2]:
archivo="archivo.txt"
with open (archivo,'r') as file:
    lineas = file.readlines()

for linea in lineas:
    print(linea.strip())
"""Esribe tu código aquí"""

1 2
2 4
3 6
4 8


'Esribe tu código aquí'

y haz que tenga la siguiente salida:

    1.0   2.00  
    2.0   4.00  
    3.0   6.00  
    4.0   8.00  


<hr>
<b><font color='red'>Ejercicio 02</font></b>  

Haz lo mismo con el archvo archivo_error.txt,con ciddado de que a este archivo le falta algún dato, haz que en ese caso cuando lo lea, si no hay dos columnas no lo escriba.

In [8]:
"""Esribe tu código aquí"""
with open ('archivo_error.txt','r') as file:
    lineas = file.readlines()

for linea in lineas:
    columnas = linea.strip().split()
    if len(columnas)==2:
        print(linea.strip())

1 2
3 6
4 8


<hr>
<b><font color='red'>Ejercicio 03</font></b>  

Lee el archivo.txt, haz que imprima las columnas en las filas y las filas en las columnas

In [11]:
"""Esribe tu código aquí"""
with open ('archivo.txt','r') as file:
    lineas = file.readlines()
datos = []
for linea in lineas:
    datos.append(linea.strip().split())
max_columnas = 0
for fila in datos:
    max_columnas = max(max_columnas,len(fila))
traspuesta=[]
for i in range(max_columnas):
    nueva_fila=[]
    for j in range(len(datos)):
        if i < len(datos[j]):
            nueva_fila.append(datos[j][i])
        else:
            nueva_fila.append(' ')
    traspuesta.append(nueva_fila)

for fila in traspuesta:
    print(' '.join(fila))

    

1 2 3 4
2 4 6 8


1. Escribe en un archivo la siguiente matriz:
$
\begin{equation}
\begin{pmatrix}
3 & 1 & 3\\
-1 & 1 & 0\\
-2 & 4 & 1
\end{pmatrix}
\end{equation} 
$
2. Lee el archivo
3. Calcula con los datos leidos el determinante de la matriz y lo saque por pantalla

<hr>
<b><font color='red'>Ejercicio 04</font></b>  

In [1]:
import numpy as np

with open('matriz.txt','w') as file:
    file.write('3 1 3\n')
    file.write('-1 1 0\n')
    file.write('-2 4 1\n')

with open ('matriz.txt','r') as file:
    lineas = file.readlines()

matriz = []

for linea in lineas:
    fila = list(map(int,linea.strip().split()))
    matriz.append(fila)

matriz_np = np.array(matriz)
determinante = np.linalg.det(matriz_np)

print(f'El determinante de la matriz es: {determinante}')
    

El determinante de la matriz es: -2.000000000000001


<hr>
<b><font color='red'>Ejercicio 05</font></b>  

En el criptoanálisis, la técnica de análisis de frecuencias consiste en el aprovechamiento de estudios sobre la frecuencia de las letras o grupos de letras en los idiomas para poder establecer hipótesis para aprovecharlas para poder descifrar un texto cifrado sin tener la clave de descifrado (romper). Es un método típico para romper cifrados clásicos. 
Haz una anailisis de frecuencias para los 10 primeros càpitulos de el libro "el_quijote.txt"

In [13]:
import re

# Leer el archivo
with open("el_quijote.txt", 'r', encoding='utf-8') as file:
    contenido = file.read()

# Extraer los 10 primeros capítulos (asumiendo que están marcados con "Capítulo")
capitulos = re.split(r'Capítulo [IVXLCDM]+', contenido)
texto = ''.join(capitulos[:11])  # Incluimos el texto antes del primer capítulo

# Limpiar el texto y convertir a mayúsculas
texto_limpio = re.sub(r'[^A-ZÑ]', '', texto.upper())

# Inicializar diccionario de frecuencias
frecuencias = {}
for letra in "ABCDEFGHIJKLMNÑOPQRSTUVWXYZ":
    frecuencias[letra] = 0

# Contar frecuencias
total_letras = 0
for letra in texto_limpio:
    frecuencias[letra] += 1
    total_letras += 1

# Calcular frecuencias relativas
for letra in frecuencias:
    frecuencias[letra] /= total_letras

# Mostrar resultados
print("Salida:")
for letra in "ABCDEFGHIJKLMNÑOPQRSTUVWXYZ":
    print(f"{letra}   {frecuencias[letra]:.3f}")
    
    

#for i in abc:
#    print("%.2c  " % i,"%.3f" % f[(abc.index(i))])

Salida:
A   0.127
B   0.017
C   0.037
D   0.053
E   0.138
F   0.005
G   0.011
H   0.012
I   0.053
J   0.007
K   0.000
L   0.060
M   0.026
N   0.066
Ñ   0.000
O   0.099
P   0.021
Q   0.019
R   0.061
S   0.073
T   0.037
U   0.048
V   0.011
W   0.000
X   0.000
Y   0.015
Z   0.004


Salida:

    A   0.127
    B   0.017
    C   0.037
    D   0.053
    E   0.138
    F   0.005
    .
    .
    .

<hr>
<b><font color='red'>Ejercicio 06</font></b>  

haz un analisis de frecuencias en el libro encriptado "RH RQCEPESEFM AR MPM.txt", y comparala con el analisis anterior


In [5]:
abc="ABCDEFGHIJKLMNÑOPQRSTUVWXYZ"
archivo="RH RQCEPESEFM AR MPM.txt"

"""Esribe tu código aquí"""
    
#print("el_quijote.txt","  RH RQCEPESEFM AR MPM.txt")    
#for i in abc:
#    print("%.2c  " % i,"%.3f" % f[(abc.index(i))],"       %.2c  " % i,"%.3f" % f2[(abc.index(i))])

'Esribe tu código aquí'

salida:

    el_quijote.txt   RH RQCEPESEFM AR MPM.txt
    A   0.127        A   0.049
    B   0.017        B   0.006
    C   0.037        C   0.040
    .                .
    .                .
    .                .

<hr>
<b><font color='red'>Ejercicio 07</font></b>  

Comparando el analisis de frecuencia sabrías decir cual es el titulo del libro.  
RH RQCEPESEFM AR MPM